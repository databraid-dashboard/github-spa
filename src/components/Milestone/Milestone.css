
.space-milestones {
  position: relative;
  padding-top: 7px !important;
}
