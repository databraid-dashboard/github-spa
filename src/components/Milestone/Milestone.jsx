import React from 'react';
import { Card, List, Header } from 'semantic-ui-react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ProgressArc from '../ProgressArc/ProgressArc';
import './Milestone.css';

export const Milestone = ({ title, due, created, percentComplete }) => {
  function convertDate(ISOdate) {
    const date = new Date(ISOdate);
    const formatOptions = {
      day: '2-digit',
      month: '2-digit',
      year: 'numeric',
      // hour: '2-digit',
      // minute: '2-digit',
      // hour12: true
    };

    let dateString = date.toLocaleDateString('en-US', formatOptions);
    dateString = dateString.replace(',', '')
      .replace('PM', 'p.m.')
      .replace('AM', 'a.m.');
    return dateString;
  }

  function getDueDate(date) {
    if (date === null) {
      return <div />;
    }
    return <div><Header sub>Date Due:</Header> {convertDate(date)}</div>;
  }

  return (
    <Card.Content>
      <List divided relaxed>
        <List.Item>
          <List.Content className="ui center aligned">
            <List.Header>{title}</List.Header>
            <Header sub>Date Created:</Header>
            {convertDate(created)}
            {getDueDate(due)}
            <Header sub className="space">Percent Complete:</Header>
            <ProgressArc percentComplete={percentComplete} duration={2000} />
          </List.Content>
        </List.Item>
      </List>
    </Card.Content>
  );
};

export const mapStateToProps = (state, { milestoneId }) => {
  const milestone = state.milestones.milestonesById[milestoneId];
  const { title, due, created, percentComplete } = milestone;
  return {
    title,
    due,
    created,
    percentComplete,
  };
};

Milestone.propTypes = {
  title: PropTypes.string.isRequired,
  due: PropTypes.string.isRequired,
  created: PropTypes.string.isRequired,
  percentComplete: PropTypes.number.isRequired,
};

export default connect(
  mapStateToProps,
)(Milestone);
