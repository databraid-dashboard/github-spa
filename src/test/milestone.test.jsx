import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { Milestone, mapStateToProps } from '../components/Milestone/Milestone';

const state = {
  widgets: {
    byId: {
      github: {
        milestones: {
          ids: [2601714],
          milestonesById: {
            2601714: {
              created: '2017-06-23T22:45:34Z',
              labels: ['sprint 1'],
              repoIssueNumber: 36,
              title: 'building paths and auth',
              due: '2018-06-23T22:45:34Z',
              percentComplete: 0.4,
            },
          },
        },
        loadingMilestones: false,
      },
    },
  },
};

describe('Milestone entry component', () => {
  it('should render a component with props as specified ', () => {
    const component = shallow(
      <Milestone
        title={'sprint 1'}
        due={'2018-06-23T22:45:34Z'}
        milestoneId={2601714}
        percentComplete={0.4}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('map\'s given State To Props ', () => {
    const expected = {
      title: 'building paths and auth',
      due: '2018-06-23T22:45:34Z',
      percentComplete: 0.4,
      milestoneId: 2601714,
    };
    expect(mapStateToProps(state, { milestoneId: 2601714, widgetId: 'github' })).toEqual(expected);
  });

  it('Should have a List Content aligned in the center', () => {
    const component = shallow(
      <Milestone
        title={'sprint 1'}
        due={'2018-06-23T22:45:34Z'}
        percentComplete={0.4}
        milestoneId={2601714}
      />,
    );
    expect(component.find('.aligned').exists()).toBe(true);
  });
});
