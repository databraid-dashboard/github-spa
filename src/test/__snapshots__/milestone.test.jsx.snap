// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Milestone entry component should render a component with props as specified  1`] = `
<CardContent>
  <List
    divided={true}
    relaxed={true}
  >
    <ListItem>
      <ListContent
        className="ui center aligned"
      >
        <ListHeader>
          sprint 1
        </ListHeader>
        <div>
          <Header
            className="space-milestones"
            sub={true}
          >
            Date Due: 
          </Header>
          06/23/2018
        </div>
        <Header
          className="space-milestones"
          sub={true}
        >
          Percent Complete: 
        </Header>
        <Wrapper
          duration={2000}
          milestoneId={2601714}
          percentComplete={0.4}
        />
      </ListContent>
    </ListItem>
  </List>
</CardContent>
`;
